Positions from functors

GiudGiud opened the pull request:
Vincent wants to try positions out with functions

Functors will provide this plus potential future needs (pps, scalar variables, who knows)

GiudGiud committed:
Add a Positions object using functors
refs #23587

GiudGiud committed:
Add test for position from functors
refs #23587

vincentlaboure commented on framework/src/positions/FunctorPositions.C:
```diff
@@ -0,0 +1,75 @@
+    _pos_functors.push_back(&getFunctor<Real>(name));
+
+  // Obtain the positions by evaluating the functors
+  initialize();
```
During restart, do the positions get restarted or are they evaluated based on the initial evaluation of the functor?

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1736592/) on 585d601 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>f11ac0</th>
      <th colspan=4>#25349 585d60</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.65%</td>
      <td>85.65%</td>
      <td>+0.00%</td>
      <td>94.44%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>90372</td>
      <td>90407</td>
      <td>+35</td>
      <td>34</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15140</td>
      <td>15141</td>
      <td>+1</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25349/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25349/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25349/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25349/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25349/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25349/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25349/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25349/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25349/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25349/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25349/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25349/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/25349/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25349/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25349/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25349/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25349/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25349/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25349/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25349/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25349/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25349/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25349/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25349/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25349/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25349/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25349/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25349/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25349/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25349/coverage/xfem)

</details>



This comment will be updated on new commits.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1736549/) on 585d601 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25349/site)

This comment will be updated on new commits.

lindsayad commented on framework/src/positions/FunctorPositions.C:
```diff
@@ -0,0 +1,75 @@
+  : Positions(parameters), NonADFunctorInterface(this)
+
+{
+  const auto functor_names = getParam<std::vector<MooseFunctorName>>("positions_functors");
```
```suggestion
  const auto & functor_names = getParam<std::vector<MooseFunctorName>>("positions_functors");
```

lindsayad commented on framework/src/positions/FunctorPositions.C:
```diff
@@ -0,0 +1,75 @@
+  // Locate the origin on the mesh
+  Point p(0, 0, 0);
+  auto pl = _fe_problem.mesh().getMesh().sub_point_locator();
+  auto * elem = (*pl)(p);
```
```suggestion
  auto * const elem = (*pl)(p);
```

lindsayad commented on framework/src/positions/FunctorPositions.C:
```diff
@@ -0,0 +1,75 @@
+FunctorPositions::initialize()
+{
+  clearPositions();
+  _positions.resize(_pos_functors.size() / 3);
```
```suggestion
  const auto n_positions = _pos_functors.size() / 3;
  _positions.resize(n_positions);
```

lindsayad commented on framework/src/positions/FunctorPositions.C:
```diff
@@ -0,0 +1,75 @@
+  Moose::ElemArg elem_origin = makeElemArg(elem);
+  const auto t = determineState();
+
+  for (auto i : make_range(_pos_functors.size() / 3))
```
```suggestion
  for (auto i : make_range(n_positions))
```

lindsayad commented on framework/src/positions/FunctorPositions.C:
```diff
@@ -0,0 +1,75 @@
+  auto * elem = (*pl)(p);
+  if (!elem)
+    mooseError("Origin point not in local mesh, cannot evaluate the functor there");
+  Moose::ElemArg elem_origin = makeElemArg(elem);
```
So this may not lead to an evaluation at the origin. In your `functor_positions.i` test, this would evaluate at '0.5 0.5 0'. You could create an `ElemPointArg` if you truly want to evaluate at the origin

GiudGiud replied:
changed.

GiudGiud commented on framework/src/positions/FunctorPositions.C:
```diff
@@ -0,0 +1,75 @@
+  auto * elem = (*pl)(p);
+  if (!elem)
+    mooseError("Origin point not in local mesh, cannot evaluate the functor there");
+  Moose::ElemArg elem_origin = makeElemArg(elem);
```
```suggestion
  Moose::ElemPointArg elem_origin = {elem, Point(0, 0, 0), false};
```

GiudGiud committed:
Address Alex's review

- truly evaluate at the origin
- more constness

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

Apply suggestions from code review

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

lindsayad commented on framework/src/positions/FunctorPositions.C:
```diff
@@ -0,0 +1,76 @@
+  // Use the mesh center as a global argument for now
+  _fe_problem.mesh().errorIfDistributedMesh(type());
+  // Locate the origin on the mesh
+  Point p(0, 0, 0);
```
```suggestion
  const Point p(0, 0, 0);
```

lindsayad commented on framework/src/positions/FunctorPositions.C:
```diff
@@ -0,0 +1,76 @@
+  auto * const elem = (*pl)(p);
+  if (!elem)
+    mooseError("Origin point not in local mesh, cannot evaluate the functor there");
+  Moose::ElemPointArg elem_origin = {elem, Point(0, 0, 0), false};
```
```suggestion
  Moose::ElemPointArg elem_origin = {elem, p, false};
```

GiudGiud commented:
squashed

GiudGiud commented:
Thanks for the quick review!
